### DISCUSSION TITLE
180 degrees of separation

### DISCUSSION THREAD

**Discussion by lucas-wilkins** [SasView/sasview#2421] [General]
Keeping calculation tools, data reading, guis and models conceptually separate is good.
Modularity is good, and having the ability to choose different front ends for a common API is desirable.
The components should be written so that **if** they were split into separate packages, they would function perfectly well. 
This, however, is different from keeping them in different repositories.
There is no reason why releases cannot be made of the individual components from one main repository.

Once separated by version control, the relationship between components becomes ossified - held in place by the need for changes to be made simultaneously. In software engineering terms: the contract between objects becomes fixed. 

Before even thinking about splitting components into separate respositories, a certain degree of stability needs to present. You have to be able to say: "this is what our interfaces look like, and it will be fine if they stay like this for the forseeable future. In modern development, no code is ever really stable, and there is a constant battle against bit rot. Even very seasoned, low level libraries like numpy update with breaking changes, meaning that code needs to be maintained, even if it was perfect to begin with. Keeping multiple repositories hinders the agility needed to keep on top of it.

Now, to some specifics regarding the SasView codebase...

* There is still quite a bit of work to be done on finding the right way for the components to interface. As it stands it will be relatively hard to iterate through different ideas. The more time that goes on, the more tension there will be between the different components, i.e. more compromises will be needed in the components to respect the interface.
* Issue tracking: there are lots of different places to look for issues.
* The docs: wouldn't it be good if there was just one way that the docs worked? And seriously, has anyone tried to build them!
* Maintainability: A lot has changed in the world of python. SasModels is quite behind. Updating things will require a lot of changes that will affect the other components. The longer its left, the worse it will get.
* Development environments: its currently very easy to end up with a development environment that doesn't match the compiled one.

Basically, the code is not separated enough to work in different repositories efficiently, and the version control and documentation is too separate.

**TL;DR: Modular programming good, modular version control less good. We should be thinking about having fewer repositories, not more**

**Comment by lucas-wilkins** [comment]
Here's another thing, because we pull in code from bumps/sasmodels/sasdata master branches, its possible to break the main branch without going through the CI. If you push something to sasmodels that breaks sasview, this won't get caught, because the tests that should show it are triggered by actions in sasview.

In other words, we don't treat them like different, separately versioned repositories anyway.
